src/edwards_point.rs

```diff
@@ -0,0 +1,167 @@
+    pub T: LinearCombination,
+}
+
+impl SonnyEdwardsPointGadget {
+    pub fn add(
+        self,
+        cs: &mut dyn ConstraintSystem,
+        other: SonnyEdwardsPointGadget,
+    ) -> SonnyEdwardsPointGadget {
+        // XXX: public constants should be defined at a higher level
+        let a: Scalar = Scalar::from_bytes_mod_order(zerocaf::constants::EDWARDS_A.to_bytes());
+        let d: Scalar = Scalar::from_bytes_mod_order(zerocaf::constants::EDWARDS_D.to_bytes());
+
+        // Point addition impl
+        // A = p1_x * p2_x
+        // B = p1_y * p2_y
+        // C = d*(p1_t * p2_t)
+        // D = p1_z * p2_z
+        // E = (p1_x + p1_y) * (p2_x + p2_y) + a*A + a*B
+        // F = D - C
+        // G = D + C
+        // H = B + A
+        // X3 = E * F , Y3 = G * H, Z3 = F * G, T3 = E * H
+        //
+        // Compute A
+        let (_, _, A) = cs.multiply(self.X.clone(), other.X.clone());
+
+        // Compute B
+        let (_, _, B) = cs.multiply(self.Y.clone(), other.Y.clone());
+
+        // Compute C
+        let (_, _, pt) = cs.multiply(self.T, other.T);
+        let (_, _, C) = cs.multiply(pt.into(), d.into());
+
+        // Compute D
+        let (_, _, D) = cs.multiply(self.Z, other.Z);
+
+        // Compute E
+        let E = {
+            let E1 = self.X.clone() + self.Y.clone();
+            cs.constrain(E1.clone() - self.X - self.Y);
+
+            let E2 = other.X.clone() + other.Y.clone();
+            cs.constrain(E2.clone() - other.X - other.Y);
+
+            let (_, _, E12) = cs.multiply(E1, E2);
+
+            let (_, _, aA) = cs.multiply(a.into(), A.into());
+            let (_, _, bB) = cs.multiply(a.into(), B.into());
+
+            let E = aA + bB + E12;
+            cs.constrain(E.clone() - aA - bB - E12);
+
+            E
+        };
+
+        // Compute F
+        let F = D - C;
+        cs.constrain(F.clone() - D + C);
+
+        // Compute G
+        let G = D + C;
+        cs.constrain(G.clone() - D - C);
+
+        // Compute H
+        let H = B + A;
+        cs.constrain(H.clone() - B - A);
+
+        // Compute resulting point
+        let (E, F, X) = cs.multiply(E, F);
+        let (G, H, Y) = cs.multiply(G, H);
+        let (_, _, Z) = cs.multiply(F.into(), G.into());
+        let (_, _, T) = cs.multiply(E.into(), H.into());
+
+        SonnyEdwardsPointGadget {
+            X: X.into(),
+            Y: Y.into(),
+            Z: Z.into(),
+            T: T.into(),
+        }
+    }
+    // self.x * other.z = other.x * self.z AND self.y * other.z == other.y * self.z
+    pub fn equal(&self, cs: &mut dyn ConstraintSystem, other: SonnyEdwardsPointGadget) {
+        let (_, _, a) = cs.multiply(self.X.clone(), other.Z.clone());
+        let (_, _, b) = cs.multiply(other.X, self.Z.clone());
+        cs.constrain(a - b);
+
+        let (_, _, c) = cs.multiply(self.Y.clone(), other.Z);
+        let (_, _, d) = cs.multiply(other.Y, self.Z.clone());
+        cs.constrain(c - d);
+    }
```

Comment:
If this is comparing two `RistrettoPoint`s, the formulae is not correct, it should be done following the [Ristretto formulas](https://ristretto.group/details/equality.htmll). Check the last equation to see it.

Reply:
I removed RistrettoPoint because points are de-facto Twisted Edwards, AFAIK we cannot be sure that they are in the Ristretto subgroup

Reply:
We can if we only accept `RistrettoPoint` types in the input values phase. Otherways, it will be difficult to check inside of the circuit.